contracts/Gauge.sol

```diff
@@ -261,15 +290,38 @@ contract Gauge is BaseGauge, IGauge {
         }
     }
 
+    function _beforeTokenTransfer(
+        address _from,
+        address _to,
+        uint256
+    ) internal override {
+        if (_from != address(0)) {
+            _updateReward(_from);
+            //also deposit to linked rewards
+            uint256 length = extraRewards.length;
+            for (uint256 i = 0; i < length; ++i) {
+                IExtraReward(extraRewards[i]).rewardCheckpoint(_from);
+            }
+        }
+        if (_to != address(0)) {
+            _updateReward(_to);
```

Comment:
what about updating balance.realbalance and other struct related data in this point, it seems transfers should put it out of sync with realBalance.

Say a user deposits, later gets a transfer from a second user, balance.realbalance, balance.lastDeposit and balance.boostedBalance will be out of sync for the user, i guess not in his favor.

lastDeposit seems to make sense to update based on a transfer, basically a transfer should be equivalent to a depositFor in accounting housekeeping.

Reply:
I agree,

If I can transfer my gTokens, when I transfer it to another user, we should do something similar to what we do in the `_depositFor` and `_withdraw` functions:

- [_depositFor](https://github.com/yearn/veYFI/blob/IERC4626/contracts/Gauge.sol#L493)
- [_withdraw](https://github.com/yearn/veYFI/blob/IERC4626/contracts/Gauge.sol#L632-L634)

Essentially, we update the `_boostedBalances` for the given account, but in this case I think we should do it for both addresses (`_to` and `_from`):

```
        uint256 boostedBalance = _boostedBalanceOf(_receiver);
        _boostedBalances[_receiver] = boostedBalance;

        emit BoostedBalanceUpdated(_receiver, boostedBalance);
```

A side note: I think we should add that logic in the `_afterTokenTransfer` function, because we are using `_beforeTokenTransfer` and the account balances will not be updated at that point in time to update the `boostedBalances`. Please, CMIIW.

Reference:  The [transfer](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L226-L246) function has two hooks:

- [_beforeTokenTransfer(from, to, amount)](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L234)
- [_afterTokenTransfer(from, to, amount)](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L245)

Reply:
Also, imo we should add unit tests to cover all the potential scenarios, and test it on ropsten too.

Reply:
https://github.com/yearn/veYFI/pull/155/commits/bd5a0069001f00fb8b9df5763a05bd8057bfd8d0

Reply:
LAtest commit looks good, but needs to add the unit tests around transfer use case and reward checks, transfering several times in a loop back and forth between two accounts and see if rewards are consistent.